Pi2BoardPkg/Drivers/DisplayDxe/DisplayDxe.c

```diff
@@ -202,6 +202,81 @@ DisplaySetMode(
     return EFI_SUCCESS;
 }
 
+BOOLEAN TruncateBltRange(
+    IN UINT32 GopUpperBound,
+    IN UINTN BltOffset,
+    IN OUT UINTN *BltRange,
+    OUT EFI_STATUS *Status
+)
+{
+    *Status = EFI_SUCCESS;
+    if (BltOffset + *BltRange > GopUpperBound) {
+        if (GopUpperBound > BltOffset) {
+            *BltRange = GopUpperBound - BltOffset;
+            return TRUE;
+        } else {
+            *Status = EFI_INVALID_PARAMETER;
+        }
+    }
+    return FALSE;
+}
+
+EFI_STATUS TruncateBltRangeIfNecessary(
+    IN EFI_GRAPHICS_OUTPUT_BLT_OPERATION       BltOperation,
+    IN EFI_GRAPHICS_OUTPUT_PROTOCOL            *Gop,
+    IN UINTN                                   SourceX,
+    IN UINTN                                   SourceY,
+    IN UINTN                                   DestinationX,
+    IN UINTN                                   DestinationY,
+    IN OUT UINTN                               *Width,
+    IN OUT UINTN                               *Height
+)
+{
+    UINT32 GopWidth = Gop->Mode->Info->HorizontalResolution;
+    UINT32 GopHeight = Gop->Mode->Info->VerticalResolution;
+    EFI_STATUS Status = EFI_SUCCESS;
+
+    if (EfiBltVideoFill == BltOperation || EfiBltBufferToVideo == BltOperation || EfiBltVideoToVideo == BltOperation) {
```

Comment:
This might be easier to read using a switch statement.

Reply:
In this particular case, a switch won't make things more readable. The two if's share one case "EfiBltVideoToVideo == BltOperation". So I can't always 'break' between the corresponding switch cases.